### PR TITLE
Add API to save anomaly data

### PR DESCRIPTION
**Pull request by mikevoronov**
This PR adds two new methods in `DataStore` to determine anomaly and collect necessary data. Additionally
  - a generic parameter of `DataStore` turned to be associative
  - `AVMOutcome` contains additionally `memory_delta` and `execution_time`

### CODE REVIEW

File: avm/server/src/avm.rs [hunk 1]
```diff
@@ -81,20 +83,29 @@ impl<E> AVM<E> {
...
+                current_data.clone(),
+                particle_parameters.init_peer_id.clone().into_owned(),
                 particle_parameters.timestamp,
                 particle_parameters.ttl,
                 call_results,
             )
             .map_err(AVMError::RunnerError)?;
 
+        let execution_time = execution_start_time.elapsed();
+        let memory_delta = self.runner.memory_stats().memory_size - memory_size_before;
```
Comment (ValeryAntopol) [inline avm/server/src/avm.rs:104]:
```suggestion
        let memory_delta = self.memory_stats().memory_size - memory_size_before;
```

**Review by ValeryAntopol — COMMENTED** [review]
utACK

File: crates/data-store/src/lib.rs [hunk 2]
```diff
@@ -14,13 +14,35 @@
...
+
+    /// Cleanup data that become obsolete.
+    fn cleanup_data(&mut self, key: &str) -> Result<(), Self::Error>;
+
+    /// Returns true if an anomaly happened and it's necessary to save execution data
+    /// for debugging purposes.
+    ///  execution_time - is a time taken by the interpreter to execute provided script
+    ///  memory_delta - is a count of bytes on which an interpreter heap has been extended
+    ///                 during execution of a particle
+    fn detect_anomaly(&self, execution_time: Duration, memory_delta: usize) -> bool;
```
Comment (monoid) [inline crates/data-store/src/lib.rs:39]: It might be bit disputable to put anomaly detection logic into a DataStore.

File: avm/server/src/avm.rs [hunk 3]
```diff
@@ -81,23 +83,32 @@ impl<E> AVM<E> {
...
+        let execution_start_time = Instant::now();
+        let memory_size_before = self.memory_stats().memory_size;
         let outcome = self
             .runner
             .call(
                 air,
                 prev_data,
-                data,
-                particle_parameters.init_peer_id.into_owned(),
+                current_data.clone(),
```
Comment (monoid) [inline avm/server/src/avm.rs:95]: I have lot of ideas to discuss looking at this line.

File: crates/data-store/src/lib.rs (hunk 2 above)
Comment (mikevoronov) [reply crates/data-store/src/lib.rs:39]: I had the same thoughts and we dicsussed it with @folex and it looks like node code would become too complicated after introducing a new boxed trait to AVM. So we decided to do that in this trait.

**Review by folex — APPROVED** [review]
Reading `prev_data` from disk in `save_anomaly_data` looks a little bit error-prone: gotta be careful to not overwrite it with `new_data` before `save_anomaly_data` is called. But I understand the motivation.

So utACK :)

**Review by folex — APPROVED** [review]
Tested ACK

**Pull request merged**